### PR TITLE
Implement sgemm and dgemm using fma

### LINKED ISSUE #31: Explore performance of _mm256_blend_ps vs _mm256_shuffle_ps
While implementing the dgemm kernel, I noticed that one can choose to either a) use `_mm256_blend_ps` followed by `_mm256_permute2f128_ps` or b) use `_mm256_shuffle_ps` followed by `_mm256_permute2f128_ps` to achieve the same goal (this is at the end, when scaling the product of `a` and `b` by `alpha`, and `c` by `beta`).

Doing the first operation leads to packed simd vectors containing a column (of 8 rows) each, while doing the second operation gives rows (containing 8 columns each).

Currently, the `sgemm` kernel implements option b), where `_mm256_shuffle_ps` has latency 1 and throughput 1. Doing option a) we'd get latency 1 but throughput 0.33 (on most Intel architectures).

It's worth investigating if this improves performance.

### LINKED ISSUE #35: Use fma, fused multiply add, for architectures supporting fma
Modern Intel architectures supporting fma instruction sets can perform the first loop calculating the matrix-matrix product between panels a and b in one go using `_mm256_fmadd_pd`. We should implement these and see how they affect performance.

### LINKED ISSUE #38: Test fma instructions on travis
It looks like not all build servers travis uses support modern `fma` instruction, see for example https://github.com/bluss/matrixmultiply/pull/36#issuecomment-444258794

One workaround is to specify `os: linux`, `sudo: true`, and `dist: trusty`, which turns out to activate `fma`. The reason why is unknown, but it was first documented here: https://github.com/uclouvain/openjpeg/blob/master/.travis.yml#L29-L33 Are these images for whatever reason tied to certain servers with a modern architecture?

This is prone to fail in the future, should this trick stop working.

### PR DESCRIPTION
**Pull request by SuperFluffy**
This uses fused multiply add via `_mm256_fmadd_{ps,pd}` to multiply and accumulate matrices in one go. The performance gains are impressive, as described in issue #35.

Fixes #31 
Fixes #35 
Fixes #38

### CODE REVIEW

**Comment by bluss** [comment]
Can you update Travis and unit tests so that they still cover all kernels?

**Comment by bluss** [comment]
Let's merge the other one then we rebase and fix this pr.

**Comment by SuperFluffy** [comment]
Let's do that! I haven't yet looked into why the tests in the other one,
and hence this one, are failing.

**Comment by bluss** [comment]
This PR is failing on travis for its own reason, so it should be investigated.

**Comment by SuperFluffy** [comment]
> This PR is failing on travis for its own reason, so it should be investigated.

Could it be that it's failing due to there not being support for fma on travis?

**Comment by bluss** [comment]
@SuperFluffy The code should runtime detect if it can use fma or not, so then there is a bug. Also, are you sure the feature "fma" implies "avx"?  I haven't reviewed this, so I'm not sure but I think we need to manually check for each intrinsic if it belongs to the correct feature (in this case the fma feature).

**Comment by bluss** [comment]
This build is crashing with SIGILL, that sounds interesting. https://travis-ci.org/bluss/matrixmultiply/jobs/462848596

**Comment by bluss** [comment]
I got the travis builder to spit out its available target features and indeed it doesn't support fma. But why did it crash? And how can we keep this tested if travis doesn't have it...

```rust
> rustc --print cfg -Ctarget-cpu=native
debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="gnu"
target_family="unix"
target_feature="avx"
target_feature="fxsr"
target_feature="mmx"
target_feature="pclmulqdq"
target_feature="popcnt"
target_feature="rdrand"
target_feature="sse"
target_feature="sse2"
target_feature="sse3"
target_feature="sse4.1"
target_feature="sse4.2"
target_feature="ssse3"
target_feature="xsave"
target_feature="xsaveopt"
target_has_atomic="16"
target_has_atomic="32"
target_has_atomic="64"
target_has_atomic="8"
target_has_atomic="cas"
target_has_atomic="ptr"
target_os="linux"
target_pointer_width="64"
target_thread_local
target_vendor="unknown"
unix
```

**Comment by bluss** [comment]
Can you use github keywords to close issues? https://help.github.com/articles/closing-issues-using-keywords/  In this case, just put "Fixes #35" in the PR description. The PR description is the best place to put this.

We need to resolve the massive code duplication and comment duplication. (It's almost exactly the same code, isn't it?). Is it likely to stay identical like this?

I'd propose to solve it by making exactly `kernel_x86_avx` a generic function. Make a simple trait and two marker types, so that you can call `kernel_x86_avx::<Fma>` and `kernel_x86_avx::<Avx>`. Using conditionals aided by those static types will make us generate two distinct functions at compile time.

**Comment by SuperFluffy** [comment]
So, this feels like a bit of a hack, but I found this when googling: https://github.com/uclouvain/openjpeg/blob/master/.travis.yml#L29-L33

If you specify `os: linux`, `sudo: true`, and `dist: trusty`, you get a machine with `avx2` and `fma`, apparently. The tests pass now....

> Can you use github keywords to close issues? https://help.github.com/articles/closing-issues-using-keywords/ In this case, just put "Fixes #35" in the PR description. The PR description is the best place to put this.

Will do!

> We need to resolve the massive code duplication and comment duplication. (It's almost exactly the same code, isn't it?). Is it likely to stay identical like this?

Yes, you are right, we should fix that.

**Comment by bluss** [comment]
@SuperFluffy but tests should pass also on machines that don't have fma. I'm not sure why they were failing, can we understand that?

**Comment by SuperFluffy** [comment]
@bluss: Yes, you are right once more. Looks like the macro isn't picking up on `fma` not being available?

Regarding what you said earlier:

> Also, are you sure the feature "fma" implies "avx"?

From what I can tell, there is not a single procecessor out there that supports `fma`, but not `avx`. Since `fma` is acting on `__m256` and `__m256d` vectors, and since the only way to load them is through functions like `_mm256_load_p{s,d}` introduced with avx, I think we can safely assume that `fma => avx`.

**Comment by bluss** [comment]
The tests need to be updated so that they crash again. That they pass is indicative of one thing: We don't test the avx kernel on this new fma setup :smile: 

So .travis.yml would need to be updated to make sure we reach all the different kernels.

File: src/sgemm_kernel.rs [hunk 1]
```diff
@@ -295,6 +306,10 @@ unsafe fn kernel_x86_avx(k: usize, alpha: T, a: *const T, b: *const T,
     ab[6] = ab6666;
     ab[7] = ab7777;
 
+    macro_rules! c {
+        ($i:expr, $j:expr) => (c.offset(rsc * $i as isize + csc * $j as isize));
+    }
+
```
Comment (bluss) [inline src/sgemm_kernel.rs]: duplicated macro def

Comment (SuperFluffy) [reply src/sgemm_kernel.rs]: Forgot to force-push my fix for that last night.

**Comment by bluss** [comment]
Thanks! Massive performance improvement. Will add on the dedup of sgemm too.

**Pull request merged**